source/opt/loop_utils.cpp

```diff
@@ -0,0 +1,379 @@
+    // If non_dedicate is in a loop, add the new dedicated exit in that loop.
+    if (ir::Loop* parent_loop = loop_desc[non_dedicate])
+      parent_loop->AddBasicBlock(&exit);
+  }
+
+  if (new_loop_exits.size() == 1) {
+    loop_->SetMergeBlock(*new_loop_exits.begin());
+  }
+
+  if (made_change) {
+    context_->InvalidateAnalysesExceptFor(
+        PreservedAnalyses | ir::IRContext::kAnalysisCFG |
+        ir::IRContext::Analysis::kAnalysisLoopAnalysis);
+  }
+}
+
+void LoopUtils::MakeLoopClosedSSA() {
+  CreateLoopDedicateExits();
+
+  ir::Function* function = loop_->GetHeaderBlock()->GetParent();
+  ir::CFG& cfg = *context_->cfg();
+  opt::DominatorTree& dom_tree =
+      context_->GetDominatorAnalysis(function, cfg)->GetDomTree();
+
+  opt::analysis::DefUseManager* def_use_manager = context_->get_def_use_mgr();
+  std::unordered_set<ir::BasicBlock*> exit_bb;
+  for (uint32_t bb_id : loop_->GetBlocks()) {
+    ir::BasicBlock* bb = cfg.block(bb_id);
+    bb->ForEachSuccessorLabel([&exit_bb, &cfg, this](uint32_t succ) {
+      if (!loop_->IsInsideLoop(succ)) exit_bb.insert(cfg.block(succ));
+    });
+  }
+
+  for (uint32_t bb_id : loop_->GetBlocks()) {
+    ir::BasicBlock* bb = cfg.block(bb_id);
+    // If bb does not dominate an exit block, then it cannot have escaping defs.
+    if (!DominatesAnExit(bb, exit_bb, dom_tree)) continue;
+    for (ir::Instruction& inst : *bb) {
+      std::unordered_set<ir::BasicBlock*> processed_exit;
+      LCSSARewriter rewriter(context_, dom_tree, exit_bb, inst);
+      def_use_manager->ForEachUse(
+          &inst, [&rewriter, &exit_bb, &processed_exit, &inst, &dom_tree, &cfg,
+                  this](ir::Instruction* use, uint32_t operand_index) {
+            if (loop_->IsInsideLoop(use)) return;
+
+            ir::BasicBlock* use_parent = context_->get_instr_block(use);
+            assert(use_parent);
+            if (use->opcode() == SpvOpPhi) {
+              // If the use is a Phi instruction and the incoming block is
+              // coming from the loop, then that's consistent with LCSSA form.
+              if (exit_bb.count(use_parent)) {
+                rewriter.RegisterExitPhi(use_parent, use);
+                return;
+              } else {
+                // That's not an exit block, but the user is a phi instruction.
+                // Consider the incoming branch only: |use_parent| must be
+                // dominated by one of the exit block.
+                use_parent = context_->get_instr_block(
+                    use->GetSingleWordOperand(operand_index + 1));
+              }
+            }
+
+            for (ir::BasicBlock* e_bb : exit_bb) {
+              if (processed_exit.count(e_bb)) continue;
+              processed_exit.insert(e_bb);
+
+              // If the current exit basic block does not dominate |use| then
+              // |inst| does not escape through |e_bb|.
+              if (!dom_tree.Dominates(e_bb, use_parent)) continue;
```

Comment:
I'm not convinced this is true when the CFG is not structured.  Can you give more of an explanation?  I'm thinking a loop header defines a value, say `%2`.  It dominates its two exits nodes.  The two exit nodes then branch unconditionally to a node `bb4` that uses `%2`.  Then `bb4` is not dominated by either of the exit nodes, but `%2` still escapes through both of them.

What is suppose to happen in a case like this?

Reply:
That's a (stupid) mistake, this is true if use_parent does not post-dominates e_bb ...

I wanted to add a simple check that try to reduce the number of unnecessary phis, but by addressing the performance issue you raised for the rewriter, that can be done without that domination check.